tokio-timer/src/deadline.rs

```diff
@@ -0,0 +1,165 @@
+//! Docs
+
+use Sleep;
+
+use futures::{Future, Poll, Async};
+
+use std::time::Instant;
+
+/// Allows a given `Future` up to the specified deadline.
+///
+/// If the inner future completes before the deadline is reached, then
+/// `Deadline` completes with that value. Otherwise, `Deadline` complets with a
+/// [`DeadlineError`].
+///
+/// [`DeadlineError`]: struct.DeadlineError.html
+#[must_use = "futures do nothing unless polled"]
+#[derive(Debug)]
+pub struct Deadline<T> {
+    future: T,
+    sleep: Sleep,
+}
+
+/// Error returned by `Deadline` future.
+#[derive(Debug)]
+pub struct DeadlineError<T> {
+    kind: Kind<T>,
+}
+
+/// Deadline error variants
+#[derive(Debug)]
+enum Kind<T> {
+    /// Inner future returned an error
+    Inner(T),
+
+    /// The deadline elapsed.
+    Elapsed,
+
+    /// Timer returned an error.
+    Timer(::Error),
+}
+
+impl<T> Deadline<T> {
+    /// Create a new `Deadline` that completes when `future` completes or when
+    /// `deadline` is reached.
+    pub fn new(future: T, deadline: Instant) -> Deadline<T> {
+        Deadline::new_with_sleep(future, Sleep::new(deadline))
+    }
+
+    pub(crate) fn new_with_sleep(future: T, sleep: Sleep) -> Deadline<T> {
+        Deadline {
+            future,
+            sleep,
+        }
+    }
+
+    /// Gets a reference to the underlying future in this deadline.
+    pub fn get_ref(&self) -> &T {
+        &self.future
+    }
+
+    /// Gets a mutable reference to the underlying future in this deadline.
+    pub fn get_mut(&mut self) -> &mut T {
+        &mut self.future
+    }
+
+    /// Consumes this deadline, returning the underlying future.
+    pub fn into_inner(self) -> T {
+        self.future
+    }
+}
+
+impl<T> Future for Deadline<T>
+where T: Future,
+{
+    type Item = T::Item;
+    type Error = DeadlineError<T::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // First, try polling the future
+        match self.future.poll() {
+            Ok(Async::Ready(v)) => return Ok(Async::Ready(v)),
+            Ok(Async::NotReady) => {}
+            Err(e) => return Err(DeadlineError::inner(e)),
+        }
+
+        // Now check the timer
+        match self.sleep.poll() {
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Ok(Async::Ready(_)) => {
+                Err(DeadlineError::elapsed())
+            },
+            Err(e) => Err(DeadlineError::timer(e)),
+        }
+    }
+}
+
+// ===== impl DeadlineError =====
```

Comment:
I think this ought to provide a `std::error::Error` implementation `where T: Error` (and similar `fmt::Display` impl).